### PR TITLE
[ci] feat: add npu unit test

### PR DESCRIPTION
**Pull request by yyyy2000**
### What does this PR do?

> Add npu unit test workflow. Enable unit tests to auto-detect device (GPU/NPU)

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: tests/single_controller/test_worker_group_basics.py [hunk 1]
```diff
@@ -38,6 +39,19 @@ def two_to_all_dispatch_fn(worker_group, *args, **kwargs):
...
+    """Function that gets the torch.device based on the current machine.
+    This currently only supports CPU, CUDA, NPU.
+    Returns:
+        device
+    """
+    if torch.cuda.is_available():
+        device = "GPU"
+    elif torch.npu.is_available():
+        device = "NPU"
+    return device
```
[the remark this answers is not in the document]
Comment (yyyy2000) [reply tests/single_controller/test_worker_group_basics.py]: 已修改

File: tests/workers/actor/test_special_dp_actor.py [hunk 2]
```diff
@@ -68,6 +69,9 @@ def setUpClass(cls):
         if torch.cuda.is_available():
             torch.cuda.set_device(cls.rank)
             cls.device = torch.device(f"cuda:{cls.rank}")
+        elif torch.npu.is_available():
+            torch.npu.set_device(cls.rank)
+            cls.device = torch.device(f"npu:{cls.rank}")
         else:
             cls.device = torch.device("cpu")
```
[the remark this answers is not in the document]
Comment (yyyy2000) [reply tests/workers/actor/test_special_dp_actor.py]: 已修改

File: tests/utils/test_torch_functional.py [hunk 3]
```diff
@@ -19,21 +19,21 @@
...
-        backend="nccl",
+        backend=get_nccl_backend(),
         init_method=f"file://{rendezvous_file}",
         rank=rank,
         world_size=world_size,
     )
 
-    # each rank holds tensor [rank+1]
-    local = torch.tensor([float(rank + 1)], device=f"cuda:{rank}")
+    local = torch.tensor([float(rank + 1)], device=get_device_name())
```
[the remark this answers is not in the document]
Comment (yyyy2000) [reply tests/utils/test_torch_functional.py]: 已修改

File: .github/workflows/npu_unit_tests.yml [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+#   - Since cpu/gpu/npu unit tests by default runs all tests under `tests`, please make sure tests are manually excluded in them when
+#     - new workflow yaml is added to `.github/workflows`
+#     - new tests are added to workflow mentioned in 2.
+
+name: NPU unit tests
+
+on:
+  # Trigger the workflow on push or pull request,
+  # but only for the main branch
+  push:
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml:38]: Set specific paths for `push` branch is required, see https://github.com/volcengine/verl/blob/0c06358d6b5624fe4e98f7f003ad623289571611/.github/workflows/gpu_unit_tests.yml#L41-L43

File: .github/workflows/npu_unit_tests.yml [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  cancel-in-progress: ${{ github.ref != 'refs/heads/main' }}
+
+# Declare permissions just read content.
+permissions: 
+  contents: read
+
+jobs:
+  test:
+    if: github.repository_owner == 'volcengine'
+    name: verl Ascend test (self-host)
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml]: How about naming the job as `Ascend unit test` ?

File: .github/workflows/npu_unit_tests.yml [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/main' }}
+
+# Declare permissions just read content.
+permissions: 
+  contents: read
+
+jobs:
+  test:
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml]: How about naming sub job name as `npu_unit_tests` ?

File: .github/workflows/npu_unit_tests.yml [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      - name: Checkout volcengine/verl repo
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          clean: true
+      - name: Install the current repository
+        run: |
+          pip install -r requirements-npu.txt
+          pip install -e .
+          pip install --no-deps -e .[test]
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml]: It is not necessary to install verl twice, just install with [test] is fine

File: .github/workflows/npu_unit_tests.yml [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+          pip install pytest-asyncio
+          pip install mlflow
+          pip install triton
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml]: For Ascend NPU, just triton-ascend is required, and have already be installed with `requirements-npu.txt`, manually installing Triton is not required.

File: .github/workflows/npu_unit_tests.yml [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        with:
+          fetch-depth: 0
+          clean: true
+      - name: Install the current repository
+        run: |
+          pip install -r requirements-npu.txt
+          pip install -e .
+          pip install --no-deps -e .[test]
+          pip install pytest-asyncio
+          pip install mlflow
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml]: Combine into single line

File: .github/workflows/npu_unit_tests.yml [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+          pip install triton
+      - name: Check final pip list
+        run: |
+          pip list
+      - name: Prepare weights
+        run: |
+          hf download Qwen/Qwen2.5-0.5B-Instruct --local-dir ~/models/Qwen/Qwen2.5-0.5B-Instruct
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml]: CI must use local cached model to prevent frequent calling download api, which will cause API usage block by HuggingFace, contact Ascend developer to pre-cache target models.

File: .github/workflows/npu_unit_tests.yml [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+          hf download Qwen/Qwen2.5-1.5B-Instruct --local-dir ~/models/Qwen/Qwen2.5-1.5B-Instruct
+      - name: Run all NPU unit tests
+        run: |
+          pytest -s -x --ignore-glob="*test_special_*.py" --ignore-glob="*on_cpu.py" --ignore-glob="*test_vllm*" --ignore-glob="*_sglang*" --ignore-glob="*_hf_rollout*" --ignore-glob="tests/models/" --ignore-glob="tests/special*" --ignore-glob="tests/experimental" --ignore-glob="tests/workers/reward_model" --ignore-glob="*test_rvdz*" --ignore-glob="*test_ray_collectives*" --ignore-glob="*test_nvtx_profile*" tests/
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml:113]: Suggest moving following Ascend mstx profiling unit test case into this file, see 
https://github.com/volcengine/verl/blob/0c06358d6b5624fe4e98f7f003ad623289571611/.github/workflows/e2e_ascend.yml#L113-L116

File: .github/workflows/npu_unit_tests.yml [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    runs-on: linux-aarch64-a2-8
+    timeout-minutes: 15 # Increase this timeout value as needed
+    container:
+      image: swr.ap-southeast-1.myhuaweicloud.com/base_image/ascend-ci/verl/verl:verl-8.3.rc1-910b-ubuntu22.04-py3.11-latest
+      options: >-
+        --shm-size 16g
+    env:
+      HTTP_PROXY: ${{ secrets.PROXY_HTTP }}
+      HTTPS_PROXY: ${{ secrets.PROXY_HTTPS }}
+      NO_PROXY: "localhost,127.0.0.1,hf-mirror.com"
```
Comment (FightingZhen) [inline .github/workflows/npu_unit_tests.yml]: PROXY related envs are not requried any more, please remove them

File: tests/single_controller/test_worker_group_torch.py [hunk 13]
```diff
@@ -68,13 +69,15 @@ def test_all_gather_torch():
     """
     In this test, we instantiate 4 GPUs in a group and test the all_gather
     """
-    ray.init()
+    ray.init(num_cpus=100)
```
Comment (FightingZhen) [inline tests/single_controller/test_worker_group_torch.py]: Why set 100 here? same as follows ~

File: tests/utils/test_activation_offload.py [hunk 14]
```diff
@@ -26,15 +26,15 @@
...
 from verl.utils.activation_offload import enable_activation_offloading
 from verl.utils.checkpoint.fsdp_checkpoint_manager import FSDPCheckpointManager
+from verl.utils.device import get_device_name, get_nccl_backend, get_torch_device
 from verl.utils.fsdp_utils import MixedPrecisionPolicy, apply_fsdp2, get_fsdp_wrap_policy
 
 
 def create_random_input_ids(batch_size, seq_len, vocab_size):
-    from flash_attn.bert_padding import unpad_input
-
+    from verl.utils.attention_utils import unpad_input
```
Comment (FightingZhen) [inline tests/utils/test_activation_offload.py]: Not advised, should consider is_cuda_available or is_npu_available. For cuda, should keep existing logic, for NPU, use `unpad_input` in `verl.utils.npu_flash_attn_utils`

File: tests/utils/test_special_linear_cross_entropy_tp.py [hunk 15]
```diff
@@ -508,7 +511,7 @@ def check_kernel_storage(self):
         if VERIFY_TORCH_SELF:
             test.verify_torch_itself()
         test.check_torch_storage()
-        test.verify_kernel_correctness()
-        test.check_kernel_storage()
+        # test.verify_kernel_correctness()
```
Comment (FightingZhen) [inline tests/utils/test_special_linear_cross_entropy_tp.py]: Why comment both ?

File: tests/workers/actor/test_special_dp_actor.py [hunk 16]
```diff
@@ -59,15 +60,18 @@ def setUpClass(cls):
         """Set up distributed environment"""
         if not torch.distributed.is_initialized():
             torch.distributed.init_process_group(
-                backend="nccl" if torch.cuda.is_available() else "gloo", init_method="env://"
+                backend=get_nccl_backend() if get_torch_device().is_available() else "gloo", init_method="env://"
```
Comment (FightingZhen) [inline tests/workers/actor/test_special_dp_actor.py]: replace `torch.cuda.is_available()` with `is_cuda_available`

File: tests/single_controller/test_worker_group_torch.py (hunk 13 above)
Comment (yyyy2000) [reply tests/single_controller/test_worker_group_torch.py]: 100 is the original code default setting, and this will be removed

File: tests/workers/actor/test_special_dp_actor.py (hunk 16 above)
Comment (yyyy2000) [reply tests/workers/actor/test_special_dp_actor.py]: here is **get_torch_device().is_available()**   
 whether this need to be replace ？
